File: batch-expired-past-24h-window/python/openai_batch_expiry_audit.py
"""Report OpenAI batches that expired, and the ones about to.

Read only. GET requests and nothing else: give this a project key set to Read
Only. The repair is printed, never performed, because re-submitting the rows
that never ran means spending money on inference.
"""
import argparse
import logging
import os
import sys
import time

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("openai_batch_expiry_audit")

API = "https://api.openai.com/v1"

# completion_window accepts one value. This is not a default, it is the value.
WINDOW = 86400

IN_FLIGHT = ("validating", "in_progress", "finalizing", "cancelling")

# Terminal and not this note: a completed batch finished, a failed one never
# started, a cancelled one was stopped on purpose.
SETTLED = ("completed", "failed", "cancelled")

FINDINGS = ("expired", "overdue", "expiring-soon")


def counts_of(batch):
    """Read request_counts into (total, completed), or None. Pure."""
    counts = batch.get("request_counts")
    if not isinstance(counts, dict):
        return None
    try:
        return (int(counts.get("total") or 0), int(counts.get("completed") or 0))
    except (TypeError, ValueError):
        return None


def deadline(batch):
    """When this batch's window closes, and where the number came from. Pure.

    Returns (unix_seconds, source) or (None, reason). Three timestamps can
    answer this and they are not equally good, which is why the source is
    returned alongside the number rather than thrown away:

      expires_at      the API's own answer. Use it whenever it is there.
      in_progress_at  the window runs from when processing started, so this
                      plus 24h is the deadline whenever expires_at is absent.
      created_at      an upper bound only. Time spent in validating is not part
                      of the window, so this over-estimates the time left.
    """
    for field, offset, source in (
            ("expires_at", 0, "expires_at"),
            ("in_progress_at", WINDOW, "in_progress_at plus 24h"),
            ("created_at", WINDOW,
             "created_at plus 24h, an upper bound: the window starts when the "
             "batch starts processing, not when it was created")):
        raw = batch.get(field)
        if raw in (None, ""):
            continue
        try:
            value = int(raw)
        except (TypeError, ValueError):
            continue
        if value > 0:
            return (value + offset, source)
    return (None, "no usable timestamp on this object")


def verdict(batch, now, warn_hours=4):
    """Classify one object from GET /v1/batches against a clock you pass in.

    Pure. warn_hours is the headroom below which an in-flight batch is called
    out: 4 hours left of a 24 hour window is the 20 hour mark. Returns
    (state, detail).
    """
    status = str(batch.get("status") or "").strip().lower()
    numbers = counts_of(batch)
    total, done = numbers if numbers else (0, 0)
    rows = ("%d of %d row(s)" % (done, total)) if total else "an unreadable count of rows"

    if status == "expired":
        missing = max(0, total - done)
        return ("expired",
                "the 24 hour window closed with %d row(s) unfinished (%s done). "
                "Each one is a batch_expired line in the error file, and none of "
                "them will run." % (missing, rows))
    if status in SETTLED:
        return ("settled",
                "status is %s, so no window is running against it" % status)
    if status not in IN_FLIGHT:
        return ("unreadable",
                "status is %r, which is not a lifecycle state this script "
                "recognises" % (status or None,))

    when, source = deadline(batch)
    if when is None:
        return ("unreadable",
                "still %s and there is %s, so the window cannot be measured"
                % (status, source))

    left = when - int(now)
    hours = abs(left) / 3600.0
    if left <= 0:
        return ("overdue",
                "still %s, %.1f hour(s) past the close of its window (from %s). "
                "The rows that have not run are not going to." % (status, hours, source))
    if left <= warn_hours * 3600:
        return ("expiring-soon",
                "%.1f hour(s) of window left (from %s) with %s done. Submit the "
                "tail as a second batch while there is still time."
                % (hours, source, rows))
    return ("in-flight",
            "%.1f hour(s) of window left (from %s); %s done" % (hours, source, rows))


def get(session, path, params=None):
    r = session.get(API + path, params=params or {}, timeout=60)
    if r.status_code == 401:
        raise SystemExit("401 from OpenAI: the key is wrong, revoked, or belongs "
                         "to another project")
    r.raise_for_status()
    return r.json()


def batches(session, page_size, max_pages):
    """Walk GET /v1/batches, which paginates on the id of the last object."""
    params = {"limit": page_size}
    for _ in range(max_pages):
        page = get(session, "/batches", params)
        data = page.get("data") or []
        for batch in data:
            yield batch
        if not page.get("has_more") or not data:
            return
        params = {"limit": page_size, "after": data[-1].get("id")}


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--warn-hours", type=float, default=4.0,
                    help="call out in-flight batches with less than this many "
                         "hours of window left (default 4, the 20 hour mark)")
    ap.add_argument("--limit", type=int, default=100,
                    help="page size for GET /v1/batches (default 100)")
    ap.add_argument("--pages", type=int, default=20,
                    help="stop after this many pages (default 20)")
    ap.add_argument("--show-all", action="store_true",
                    help="also print settled batches")
    args = ap.parse_args()

    key = os.environ.get("OPENAI_API_KEY")
    if not key:
        log.error("set OPENAI_API_KEY (a project key set to Read Only)")
        return 2

    session = requests.Session()
    session.headers.update({"Authorization": "Bearer " + key})

    now = int(time.time())
    checked = 0
    expired = 0
    closing = 0
    for batch in batches(session, args.limit, args.pages):
        state, detail = verdict(batch, now, args.warn_hours)
        batch_id = str(batch.get("id") or "?")
        line = "%-15s %s  %s" % (state, batch_id, detail)
        checked += 1

        if state == "expired":
            expired += 1
            log.warning(line)
            error_file = batch.get("error_file_id")
            log.warning("  repair: rebuild a .jsonl of the custom_ids whose "
                        "error.code is batch_expired%s and re-submit them, then "
                        "split future jobs so one batch stays well under 50,000 "
                        "requests",
                        (" from GET /v1/files/%s/content" % error_file)
                        if error_file else "")
        elif state in ("overdue", "expiring-soon"):
            closing += 1
            log.warning(line)
            log.warning("  repair: store expires_at in your own job table and "
                        "alert at the 20 hour mark; a poller that waits for "
                        "status == completed waits forever on an expired batch")
        elif state == "unreadable":
            log.warning(line)
        elif args.show_all or state == "in-flight":
            log.info(line)

    log.info("%d batch(es) checked, %d expired, %d close to expiring",
             checked, expired, closing)
    return 1 if (expired or closing) else 0


if __name__ == "__main__":
    sys.exit(main())

File: batch-expired-past-24h-window/node/openai-batch-expiry-audit.mjs
/**
 * Report OpenAI batches that expired, and the ones about to.
 *
 * Read only. GET requests and nothing else: give this a project key set to Read
 * Only. The repair is printed, never performed.
 */
const API = 'https://api.openai.com/v1';

// completion_window accepts one value. This is not a default, it is the value.
const WINDOW = 86400;

const IN_FLIGHT = ['validating', 'in_progress', 'finalizing', 'cancelling'];

// Terminal and not this note.
const SETTLED = ['completed', 'failed', 'cancelled'];

/** Read request_counts into [total, completed], or null. Pure. */
export function countsOf(batch) {
  const counts = batch.request_counts;
  if (counts === null || typeof counts !== 'object' || Array.isArray(counts)) return null;
  const total = Number(counts.total ?? 0);
  const done = Number(counts.completed ?? 0);
  if (!Number.isFinite(total) || !Number.isFinite(done)) return null;
  return [Math.trunc(total), Math.trunc(done)];
}

/**
 * When this batch's window closes, and where the number came from. Pure.
 * Returns [unixSeconds, source] or [null, reason]. Three timestamps can answer
 * this and they are not equally good, which is why the source is returned
 * alongside the number: expires_at is the API's own answer, in_progress_at plus
 * 24h is the deadline when it is absent, and created_at plus 24h is an upper
 * bound only, because time spent validating is not part of the window.
 */
export function deadline(batch) {
  const candidates = [
    ['expires_at', 0, 'expires_at'],
    ['in_progress_at', WINDOW, 'in_progress_at plus 24h'],
    ['created_at', WINDOW,
      'created_at plus 24h, an upper bound: the window starts when the batch ' +
      'starts processing, not when it was created'],
  ];
  for (const [field, offset, source] of candidates) {
    const raw = batch[field];
    if (raw === null || raw === undefined || raw === '') continue;
    const value = Number(raw);
    if (Number.isFinite(value) && value > 0) return [Math.trunc(value) + offset, source];
  }
  return [null, 'no usable timestamp on this object'];
}

/**
 * Classify one object from GET /v1/batches against a clock you pass in. Pure.
 * warnHours is the headroom below which an in-flight batch is called out: 4
 * hours left of a 24 hour window is the 20 hour mark. Returns [state, detail].
 */
export function verdict(batch, now, warnHours = 4) {
  const status = String(batch.status ?? '').trim().toLowerCase();
  const numbers = countsOf(batch);
  const [total, done] = numbers ?? [0, 0];
  const rows = total ? `${done} of ${total} row(s)` : 'an unreadable count of rows';

  if (status === 'expired') {
    const missing = Math.max(0, total - done);
    return ['expired',
      `the 24 hour window closed with ${missing} row(s) unfinished (${rows} ` +
      'done). Each one is a batch_expired line in the error file, and none of ' +
      'them will run.'];
  }
  if (SETTLED.includes(status)) {
    return ['settled', `status is ${status}, so no window is running against it`];
  }
  if (!IN_FLIGHT.includes(status)) {
    return ['unreadable',
      `status is ${JSON.stringify(status || null)}, which is not a lifecycle ` +
      'state this script recognises'];
  }

  const [when, source] = deadline(batch);
  if (when === null) {
    return ['unreadable',
      `still ${status} and there is ${source}, so the window cannot be measured`];
  }

  const left = when - Math.trunc(now);
  const hours = (Math.abs(left) / 3600).toFixed(1);
  if (left <= 0) {
    return ['overdue',
      `still ${status}, ${hours} hour(s) past the close of its window (from ` +
      `${source}). The rows that have not run are not going to.`];
  }
  if (left <= warnHours * 3600) {
    return ['expiring-soon',
      `${hours} hour(s) of window left (from ${source}) with ${rows} done. ` +
      'Submit the tail as a second batch while there is still time.'];
  }
  return ['in-flight',
    `${hours} hour(s) of window left (from ${source}); ${rows} done`];
}

async function get(key, path, params = {}) {
  const url = new URL(API + path);
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null) url.searchParams.set(k, String(v));
  }
  const res = await fetch(url, { headers: { Authorization: `Bearer ${key}` } });
  if (res.status === 401) {
    throw new Error('401 from OpenAI: the key is wrong, revoked, or belongs to ' +
                    'another project');
  }
  if (!res.ok) throw new Error(`${res.status} from ${path}`);
  return res.json();
}

async function* walk(key, pageSize, maxPages) {
  let params = { limit: pageSize };
  for (let i = 0; i < maxPages; i += 1) {
    const page = await get(key, '/batches', params);
    const data = page.data ?? [];
    for (const batch of data) yield batch;
    if (!page.has_more || data.length === 0) return;
    params = { limit: pageSize, after: data[data.length - 1].id };
  }
}

async function main() {
  const key = (process.env.OPENAI_API_KEY || "dummy-openai-api-key");
  if (!key) {
    console.error('set OPENAI_API_KEY (a project key set to Read Only)');
    process.exitCode = 2;
    return;
  }

  const warnHours = Number((process.env.WARN_HOURS || "dummy-warn-hours") ?? 4);
  const pageSize = Number((process.env.LIMIT || "dummy-limit") ?? 100);
  const maxPages = Number((process.env.PAGES || "dummy-pages") ?? 20);
  const showAll = process.argv.includes('--show-all');
  const now = Math.floor(Date.now() / 1000);

  let checked = 0;
  let expired = 0;
  let closing = 0;
  for await (const batch of walk(key, pageSize, maxPages)) {
    const [state, detail] = verdict(batch, now, warnHours);
    const line = `${state.padEnd(15)} ${String(batch.id ?? '?')}  ${detail}`;
    checked += 1;

    if (state === 'expired') {
      expired += 1;
      console.warn(line);
      const errorFile = batch.error_file_id;
      console.warn('  repair: rebuild a .jsonl of the custom_ids whose ' +
        'error.code is batch_expired' +
        (errorFile ? ` from GET /v1/files/${errorFile}/content` : '') +
        ' and re-submit them, then split future jobs so one batch stays well ' +
        'under 50,000 requests');
    } else if (state === 'overdue' || state === 'expiring-soon') {
      closing += 1;
      console.warn(line);
      console.warn('  repair: store expires_at in your own job table and alert ' +
        'at the 20 hour mark; a poller that waits for status == completed waits ' +
        'forever on an expired batch');
    } else if (state === 'unreadable') {
      console.warn(line);
    } else if (showAll || state === 'in-flight') {
      console.log(line);
    }
  }

  console.log(`${checked} batch(es) checked, ${expired} expired, ${closing} ` +
              'close to expiring');
  process.exitCode = (expired || closing) ? 1 : 0;
}

// Only run when invoked directly, so importing this module from the test file
// does not fire main() and fail on the missing key.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
